x/ibc/callbacks/keeper/keeper.go

```diff
@@ -0,0 +1,429 @@
+	tokenPair, found := k.erc20Keeper.GetTokenPair(ctx, tokenPairID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrTokenPairNotFound, "token pair for denom %s not found", data.Token.Denom.IBCDenom())
+	}
+	amountInt, ok := math.NewIntFromString(data.Token.Amount)
+	if !ok {
+		return errorsmod.Wrapf(types.ErrNumberOverflow, "amount overflow")
+	}
+
+	erc20 := contracts.ERC20MinterBurnerDecimalsContract
+
+	remainingGas := math.NewIntFromUint64(cachedCtx.GasMeter().GasRemaining()).BigInt()
+
+	// Call the EVM with the remaining gas as the maximum gas limit.
+	// Up to now, the remaining gas is equal to the callback gas limit set by the user.
+	// NOTE: use the cached ctx for the EVM calls.
+	res, err := k.evmKeeper.CallEVM(cachedCtx, erc20.ABI, receiverHex, tokenPair.GetERC20Contract(), true, remainingGas, "approve", contractAddr, amountInt.BigInt())
+	if err != nil {
+		return errorsmod.Wrapf(types.ErrAllowanceFailed, "failed to set allowance: %v", err)
+	}
+
+	// Consume the actual used gas on the original callback context.
+	ctx.GasMeter().ConsumeGas(res.GasUsed, "callback allowance")
+	remainingGas = remainingGas.Sub(remainingGas, math.NewIntFromUint64(res.GasUsed).BigInt())
+	if ctx.GasMeter().IsOutOfGas() || remainingGas.Cmp(big.NewInt(0)) < 0 {
+		return errorsmod.Wrapf(types.ErrOutOfGas, "out of gas")
+	}
+
+	var allowance *big.Int
+	err = erc20.ABI.UnpackIntoInterface(&allowance, "allowance", res.Ret)
+	if err != nil {
+		return errorsmod.Wrapf(types.ErrAllowanceFailed, "failed to unpack allowance: %v", err)
+	}
+
+	if allowance.Cmp(big.NewInt(1)) != 0 {
+		return errorsmod.Wrapf(types.ErrAllowanceFailed, "failed to set allowance")
+	}
+
+	// NOTE: use the cached ctx for the EVM calls.
+	res, err = k.evmKeeper.CallEVMWithData(cachedCtx, receiverHex, &contractAddr, cbData.Calldata, true, remainingGas)
+	if err != nil {
+		return errorsmod.Wrapf(types.ErrEVMCallFailed, "EVM returned error: %s", err.Error())
+	}
+
+	// Consume the actual gas used on the original callback context.
+	ctx.GasMeter().ConsumeGas(res.GasUsed, "callback function")
+	if ctx.GasMeter().IsOutOfGas() {
+		return errorsmod.Wrapf(types.ErrOutOfGas, "out of gas")
+	}
+
+	// Write cachedCtx events back to ctx.
+	writeFn()
+
+	// Check that the contract has received all the tokens.
+	// NOTE: contracts must implement an IERC20(token).transferFrom(msg.sender, address(this), amount)
+	// for the total amount, or the callback will fail.
+	// This check is here to prevent funds from getting stuck in the isolated address,
+	// since they would become irretrievable.
+	contractTokenBalance := k.erc20Keeper.BalanceOf(ctx, erc20.ABI, tokenPair.GetERC20Contract(), contractAddr) // here, we can use the original ctx and skip manually adding the gas
+	if contractTokenBalance.Cmp(amountInt.BigInt()) != 0 {
+		return errorsmod.Wrapf(erc20types.ErrEVMCall, "contract balance %d does not equal sent amount %d", contractTokenBalance, amountInt.BigInt())
+	}
```

Comment:
How does this work if the callback contract already had tokens of this denom before the callback, or if the callback has already forwarded the tokens out to another account? I think you just need to check that the sender account has no tokens in it rather than assuming anything about the state of the callback contract.